Handle list

Mrod1598 opened the pull request:
**Description:**
 Resolves #42 through the proposed solution of adding the ability to pass in multiple Object Names to better support adding additional mbeans under the same metric name.

**Existing Issue(s):**
#42

**Testing:**
Added test that defines multiple mbeans and then pulls those mbeans out using multiple object names.

**Documentation:**
Documented the new signature of `otel.mbeans`

rmfitzpatrick commented:
Provided a comment on the parent issue https://github.com/open-telemetry/opentelemetry-java-contrib/issues/42#issuecomment-870981841 that could effect approach.  Overall this seems ok to me though depending on your thoughts.

Would you be willing to modify this test and its groovy script to utilize your thread pools mbean use case? https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/test/groovy/io/opentelemetry/contrib/jmxmetrics/OtlpIntegrationTests.groovy

Mrod1598 commented:
> Provided a comment on the parent issue [#42 (comment)](https://github.com/open-telemetry/opentelemetry-java-contrib/issues/42#issuecomment-870981841) that could effect approach. Overall this seems ok to me though depending on your thoughts.
> 
> Would you be willing to modify this test and its groovy script to utilize your thread pools mbean use case? https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/test/groovy/io/opentelemetry/contrib/jmxmetrics/OtlpIntegrationTests.groovy

@rmfitzpatrick I've updated the test and moved it up to the newest version. I was wondering if you could approve the CI. Thanks!

rmfitzpatrick commented on jmx-metrics/src/test/groovy/io/opentelemetry/contrib/jmxmetrics/OtlpIntegrationTests.groovy:
```diff
@@ -58,27 +60,21 @@ class OtlpIntegrationTests extends OtlpIntegrationTest {
+        Gauge datapoints = metric.gauge
         then: 'they are of the expected size'
-        datapoints.dataPointsCount == 1
+        datapoints.dataPointsCount == 44
```
I'm not sure if this is related to intermediary proto changes but the number of datapoints is now more manageable.  Would you mind adding more complete content validation?

rmfitzpatrick replied:
It may be preferable overall to keep the existing integration tests as they were and add another for this feature.

Mrod1598 replied:
moved to a separate test but had to modify the signature for `configureContainers` to allow me to pass in my own script.

Mrod1598 commented:
Altered script and test to make metrics more predictable. It was passing locally but not on actions.

Mrod1598 commented:
After testing it seems that the actions are failing through a lack of resources. The tests pass locally if docker is given enough resources but I get the same errors if I reduce the amount of memory given to docker down to 2GB. I was wondering if you had any ideas to help reduce parallelism so that the tests aren't using all of the resources available. @rmfitzpatrick

rmfitzpatrick commented:
@Mrod1598, could you try restricting the worker count for the integration test action like was done previously? https://github.com/open-telemetry/opentelemetry-java-contrib/pull/41#discussion_r660603803.  Overall better helper logic around waiting for expected metrics is needed and I will try to take care of this in the near future, but that shouldn't block this.

I think it would also be helpful if you incorporated main changes/rebased to ensure this PR is still compatible with recent updates.

Mrod1598 commented:
> @Mrod1598, could you try restricting the worker count for the integration test action like was done previously? [#41 (comment)](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/41#discussion_r660603803). Overall better helper logic around waiting for expected metrics is needed and I will try to take care of this in the near future, but that shouldn't block this.
> 
> I think it would also be helpful if you incorporated main changes/rebased to ensure this PR is still compatible with recent updates.

Sorry, forgot to mention that I tried restricting the worker count and it doesn't seem to help locally either. any other suggestions?

rmfitzpatrick commented:
@Mrod1598 thanks for your patience and I saw that it wasn't apparently effective.  I'm hoping to be able to add better waiting helpers shortly and investigate this issue.

djaglowski commented:
@rmfitzpatrick Is there anything that could be done to help move this forward?

Mrod1598 commented on jmx-metrics/src/integrationTest/resources/multi_obj_script.groovy:
```diff
@@ -0,0 +1,24 @@
+])
+otel.instrument(helper, "cassandra.current_tasks", "Number of tasks in queue with the given task status.", "1",
+        ["stage_name":{ mbean -> mbean.name().getKeyProperty("scope")},
+            "task_status":{ mbean -> mbean.name().getKeyProperty("name")}],
```
definitely have some ideas moving forward.

anuraaga commented on .github/workflows/pr-build.yml:
```diff
@@ -44,7 +44,7 @@ jobs:
         name: Integration Tests
         with:
-          arguments: --stacktrace integrationTest
+          arguments: --stacktrace --max-workers 1 integrationTest
```
Can you revert this? Don't think it's needed anymore

anuraaga commented on jmx-metrics/src/integrationTest/java/io/opentelemetry/contrib/jmxmetrics/MultiObjTest.java:
```diff
@@ -0,0 +1,73 @@
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.utility.MountableFile;
+
+class MultiObjTest extends AbstractIntegrationTest {
```
Since this looks like a new feature of the helper rather than a new backend integration I think it's better not to add yet another integration test to make sure they don't get too slow.

How about adding unit test to instrumenthelpertest and modifying the existing Cassandra test to exercise this?

Mrod1598 replied:
[here](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/43#discussion_r690663935) I had originally setup a modified version of the cassandra test and was told to revert it and add an additional test.

rmfitzpatrick replied:
My second request to keep the single mbean integration test as is and add a new one with multiples was to preserve the datapoint validation that was being removed in favor of a simple size check that was apparently flakey.

rmfitzpatrick replied:
I'm personally a fan of integration tests but can understand not want to balloon CI and dev times.  If there is to be a unit test instead of the current approach I think checking the functionality of the InstrumentHelper with multi-mbean MBeanHelper would be a good way forward.

Mrod1598 replied:
So, can this be merged? or are the changes required to be unit tests? @rmfitzpatrick

rmfitzpatrick replied:
@anuraaga are you ok with the changes as is?  Improving coverage could be a future item.

anuraaga replied:
I'd like to have the test change made yes - it's why I left a comment :P The integration tests are already up to 5 min compared to 2 for unit tests so I don't want to continue trending up.

anuraaga reviewed (commented):
@rmfitzpatrick Can you take another look at this? Would be good to verify the new groovy API

djaglowski commented:
@rmfitzpatrick Could this be merged?

Mrod1598 commented:
> @Mrod1598 thanks for updating. Would you be open to expanding the instrument helper tests a bit as mentioned earlier? [observIQ#2](https://github.com/observIQ/opentelemetry-java-contrib/pull/2)

Yes, I appreciate that

anuraaga commented:
Was hoping to merge main for you but looks like I don't have access to update the branch
